src/packages/BooleanQueryTokenizer/BooleanQueryTokenizer.ts

```diff
@@ -148,6 +155,94 @@ export class BooleanQueryTokenizer {
+                        let parenthesisTracker = 0;
+                        for (let k = j; k < tokens.length; k += 1) {
+                            const parenthesisLoopToken = tokens[k];
+                            if (parenthesisLoopToken.type === TokenType.parentheses) {
+                                if (parenthesisLoopToken.lexeme === '(') {
+                                    parenthesisTracker += 1;
+                                } else if (parenthesisLoopToken.lexeme === ')') {
+                                    parenthesisTracker -= 1;
+                                }
+                            }
+                            if (parenthesisTracker === 0) {
+                                BooleanQueryTokenizer.replaceTokens(tokens, i, k);
+                                break;
+                            }
+                        }
+                        break;
+                    } else {
+                        break;
+                    }
+                }
+            }
+        }
+        // Filter only words and phrases
+        const wordsAndPhrases = tokens
+            .filter((token) => token.type === TokenType.word || token.type === TokenType.phrase)
+            .map((wordOrPhrase) => {
+                // `phrase` tokens require some dirty preprocessing
+                // to remove quotes and possible inner wildcards.
+                if (wordOrPhrase.type === TokenType.phrase) {
+                    return wordOrPhrase.lexeme.replace(/('|"|\*)/g, '').trim();
+                }
+                return wordOrPhrase.lexeme;
+            })
+            // After `phrase` tokens preprocessing, some strings can become empty,
+            // so we should remove them.
+            .filter((str) => str);
+        return {
+            wordsAndPhrases,
+        };
+    }
+
+    private static replaceTokens(tokens: Token[], from: number, to: number): Token[] {
+        tokens.forEach((_token, index) => {
+            if (index >= from && index <= to) {
+                // eslint-disable-next-line no-param-reassign
+                tokens[index] = {
+                    type: TokenType.unknown,
+                    lexeme: '',
+                };
+            }
+        });
+        return tokens;
+    }
+
+    private static isNegation(token: Token): boolean {
+        return (
+            (token.type === TokenType.booleanOperator &&
+                (token.lexeme === 'NOT' || token.lexeme === 'AND NOT')) ||
+            (token.type === TokenType.modifier && (token.lexeme === '-' || token.lexeme === '!'))
+        );
+    }
+
+    private static isNegationable(token: Token): boolean {
+        return (
+            token.type === TokenType.word ||
+            token.type === TokenType.wildcardWord ||
+            token.type === TokenType.phrase
+        );
+    }
+
+    private static isParenthesisGroup(token: Token): boolean {
```

Comment:
I think `isParenthesisGroupStart` name is more suitable here, because we are checking exactly start of group